Step3 - 로또(자동) 입니다.

hongji3354 opened the pull request:
안녕하세요.
한 3~4일만에 PR을 올리는 것 같습니다. 처음 구현한 것이 마음에 안들어서 지우고, 포비 강의듣고 나서 코드를 다시 보니 또 마음에 안들어서 지웠다가 다시 하다 보니 오래걸렸습니다...ㅎㅎ

이번 이번 과제를 하면서 가장 신경썼던 것은 Getter 사용하지 않기 였습니다. 확실히 Getter를 사용하지 않으니 기능을 어디에 추가할지가 명확해서 좋았습니다. 
기본값 타입 변수까지 Wripping 해서 사용했으면 더 좋았을 것 같은데 그렇게 진행했다가 과제를 전혀 못끝낼 것 같아서 해당 부분은 누락되었습니다.

이번 과제를 하면서 리뷰어님께 질문 하고 싶은 것을 정리 해봤습니다. 각 질문에 대한 코드를 쉽게 확인하기 위해 다음 구조로 주석을 달아놨습니다.
- TODO : 질문1
- TODO : 질문2
- TODO : 질문3
- TODO : 질문4

## 질문1
로또를 발급하는 기능이 로또 구매 금액을 계산해서 몆 개를 구매할지의 역할을 하는 클래스에 존재하고 있습니다. 원래 LottoStore 클래스에 로또 발급 기능을 구현할려고 했는데, LottoIssuanceCount.issuanceCount를 외부에 노출을 안시키게 할려다 보니LottoIssuanceCount에 위치하게 구현하게 되었습니다. 역할 관점에서 본다면 해당 기능을  LottoIssuanceCount에 구현한 것이 좋지 않아 보이는데 데이터를 가지고 있는 곳에서 구현했다는 점을 보면 괜찮은 것 같습니다. 리뷰어님의 생각은 어떠신가요?

## 질문2
일급컬렉션인 Lottos, Lotto는 테스트를 어떻게 해야할까요?
일급컬렉션에 대한 equals를 비교할려면 내부 컬렉션과 같은 값을 컬렉션을 생성해서 비교해야 하는데 Lottos, Lotto는 Getter를 제공하지 않기 때문에 내부 컬렉션에 접근할 방법이 없습니다.
이 상황에서는 어떻게 해야할까요?

## 질문3
WinningResult.profitRate은 총 수익률을 제공하는 기능으로 구매금액만 받고 전체 당첨 금액은 내부에서 계산하는 구조 입니다. 근데 테스트시 전체 당첨 금액은 제어가 불가능 하기 떄문에 WinningResult.profitRate를 오버라이드 해서 전체 당첨 금액을 외부로 받을 수 있도록 해서 테스트를 진행하였습니다.
즉, 실제 프로덕션에서는 사용하지 않으면서 테스트를 위해 만든 것이기 때문에 좋지 않은 방법이라 생각합니다. 
어떻게 해야 테스트를 쉽게 할 수 있을 까요?

## 질문4
당첨번호의 개수를 검증하는 동일한 로직이 view와 domain에 들어가 있습니다.
view에서 검증을 하기 때문에 domain에서 빼는 것이 좋을까요? 아니면 중복이 되더라도 양쪽에서 진행을 하는 것이 좋을까요?

적다보니 질문이 많아졌네요..ㅎㅎ
리뷰어분들도 새벽에 리뷰하시는 것 같던데 고생이 많으시네요ㅠㅠ
감사합니다!

hongji3354 committed:
docs : 요구사항 최초 정리

hongji3354 committed:
feat : 로또 발급 개수 계산

hongji3354 committed:
feat : 로또 번호 생성

hongji3354 committed:
feat : 로또 구매금액을 기준으로 로또 생성

hongji3354 committed:
feat : 발급한 로또와 당첨 번호를 비교 당첨자 선정

hongji3354 committed:
feat : 당첨 정보를 가지고 수익금 및 수익률 계산

hongji3354 committed:
refactor : 코드 컨벤션

hongji3354 committed:
feat : 당첨 선별 및 당첨 금액 계산 클래스로 분리

hongji3354 committed:
refactor : domain, exception 패키지로 분리

hongji3354 committed:
refactor : 사용하지 않는 Error Message 제거

hongji3354 committed:
refactor : exception의 패키지 위치가 잘못되어 수정

hongji3354 committed:
refactor : ErrorMessage에 문자열 형식화 적용

hongji3354 committed:
fix : 구매금액이 1000원 미만일시 방어 코드 추가

hongji3354 committed:
fix : 해당 등수에 당첨자가 없을 시 0을 return 하도록 추가

hongji3354 committed:
feat : 로또 번호를 출력하기 위한 toString 재정의

hongji3354 committed:
feat : 발급한 로또 개수를 제공하는 기능

hongji3354 committed:
feat : 발급한 로또를 제공하는 기능

hongji3354 committed:
feat : 로또 당첨 번호 개수 체크하는 기능

hongji3354 committed:
refactor : Error message enum 이름이 잘못되어 수정

hongji3354 committed:
feat : view 구현

hongji3354 committed:
feat : Lotto Controller

hongji3354 committed:
style : Reformat code

hongji3354 committed:
style : 질문사항 TODO 주석

hongji3354 committed:
decs : 요구사항 완료된 것 체크

hongji3354 committed:
refactor : 전체 당첨 금액 계산하는 기능의 접근제어자 변경

leeyohan93 commented on src/main/java/lotto/domain/LottoIssuanceCount.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    //TODO : 질문1
+    public List<Lotto> createLotto() {
+        List<Lotto> lottos = new ArrayList<>();
+        for (int i = 0; i < this.issuanceCount; i++) {
+            lottos.add(new Lotto(LottoNumber.generator()));
+        }
+        return lottos;
+    }
```
> 로또를 발급하는 기능이 로또 구매 금액을 계산해서 몆 개를 구매할지의 역할을 하는 클래스에 존재하고 있습니다. 원래 LottoStore 클래스에 로또 발급 기능을 구현할려고 했는데, LottoIssuanceCount.issuanceCount를 외부에 노출을 안시키게 할려다 보니LottoIssuanceCount에 위치하게 구현하게 되었습니다. 역할 관점에서 본다면 해당 기능을 LottoIssuanceCount에 구현한 것이 좋지 않아 보이는데 데이터를 가지고 있는 곳에서 구현했다는 점을 보면 괜찮은 것 같습니다. 리뷰어님의 생각은 어떠신가요?

정말 좋은 질문입니다. 
getter를 지양해서 만들다보니 만들어진 지금의 구조이지만 저도 지훈님의 입장과 같이 만족스럽지는 못합니다. 
그 이유는 LottoIssuanceCount와 Lotto 사이에 의존성이 추가되었기 때문입니다.
LottoStore 에서 getter를 통해 count를 가져왔다면 이 의존성은 존재하지 않았겠지요. 
의존성의 추가는 높은 결합도를 의미하고, 높은 결합도는 변경하기 어려운 설계를 의미합니다.
이제 여기서 이야기의 결론으로 가보면 어떤 것을 우선해야하는 가에 대한 트레이드 오프의 시점이 왔다는 것입니다. 
두가지 모두를 만족시킬 수는 없으며 이때 저는 개인적으로 getter를 쓰더라도 좀 더 책임에 알맞아 보이는 클래스에 배치하는 것을 선호합니다 :)

추가적으로 이미 구매 금액에서 로또 개수를 구하는 책임을 가지고 있어서 로또 생성의 책임을 쥐어주기에는 너무 많은 것은 아닐까 고민해보시기도 바래요 👍

hongji3354 replied:
Getter를 사용해야 하지 않아야 한다는 목적만 달성할려다 보니 트레이드 오프 부분은 고민을 하지 못했네요.

저번주 목요일 강의에서 포비가  객체지향에서 중요한 건 객체니 객체에 집중해야 한다를고 말을 했는데, 생각해 보니 Getter를 사용하더라도 해당 책임을 가진 객체가 기능을 해주는게 맞다고 생각이 듭니다.

감사합니다!

leeyohan93 commented on src/test/java/lotto/domain/LottoStoreTest.java:
```diff
@@ -0,0 +1,16 @@
+
+class LottoStoreTest {
+
+    //TODO : 질문2
```
>일급컬렉션인 Lottos, Lotto는 테스트를 어떻게 해야할까요?
일급컬렉션에 대한 equals를 비교할려면 내부 컬렉션과 같은 값을 컬렉션을 생성해서 비교해야 하는데 Lottos, Lotto는 Getter를 제공하지 않기 때문에 내부 컬렉션에 접근할 방법이 없습니다.
이 상황에서는 어떻게 해야할까요?

일급 컬렉션에 대한 equals 비교를 하시고 싶으시다면, LottoStore를 통해 Lottos를 생성하는 것이 아니라
equals를 오버라이드 후에 생성자로 값을 직접 주입해주시면 됩니다 :)

이렇게 테스트할 수 있는 이유는 LottoStore를 통해 로또의 생성에 대한 책임을 분리해주셨기 때문이지요 👍

hongji3354 replied:
이 질문은 제가 잘못작성한 것 같습니다..ㅎㅎ

구매금액을 입력하면 로또를 제공하는 LottoStoreTest.lottoPurchase를 테스트 하고 싶은데 해당 메서드의 객체 그래프 탐색을 하다보면 로또 번호를 생성하는 LottoNumber.generator()를 만나게 됩니다.

LottoNumber.generator()는 제어가 불가능한 메서드 이기 때문에 어떻게 테스트 할수 있을지에 대한 것을 질문했어야 하는데 다른걸 질문 드렸네요.

그리고 생각 해보니 해당 테스트의 목적은 Lottos에 대한 검증이 아니라 로또 구매 금액 만큼 로또가 발급되는지 확인하는 것이 목적이 되니 LottoNumber.generator()와도 전혀 무관하게 되네요.

해당 질문에 답변하신다고 시간 쓰셨는데 죄송합니다ㅠㅠ

leeyohan93 replied:
괜찮습니다 ㅎㅎ 😄  
LottoNumber.generator()가 어려운 이유는 무엇일까요? 
답은 랜덤에 대한, 즉 테스트하기 어려운 것에 대한 의존성 때문입니다.

그렇다면 테스트하기 어려운 것에 대한 의존성을 더욱 외부로 꺼내보면 어떨까요? 🤔

leeyohan93 commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -0,0 +1,45 @@
+    //TODO 질문3
+    public double profitRate(int purchaseAmount) {
+        return profitAmount() / (double) purchaseAmount;
+    }
```
> WinningResult.profitRate은 총 수익률을 제공하는 기능으로 구매금액만 받고 전체 당첨 금액은 내부에서 계산하는 구조 입니다. 근데 테스트시 전체 당첨 금액은 제어가 불가능 하기 떄문에 WinningResult.profitRate를 오버라이드 해서 전체 당첨 금액을 외부로 받을 수 있도록 해서 테스트를 진행하였습니다.
즉, 실제 프로덕션에서는 사용하지 않으면서 테스트를 위해 만든 것이기 때문에 좋지 않은 방법이라 생각합니다.
어떻게 해야 테스트를 쉽게 할 수 있을 까요?

2번 질문과 같은 접근으로 생각해보면 굳이 protected 접근제어자를 가진 profitRate 메서드를 만드는 것 대신,
테스트에서 WinningResult 객체 생성 시 생성자를 활용해주시면 의도하신 테스트가 가능합니다 👍 
간단한 예시를 참고해주세요 :)

```java
@DisplayName("수익률 계산")
    @Test
    void profitRateTest() {
        //given
        Map<Rank, Integer> rankMap = new HashMap<>();
        rankMap.put(Rank.FOURTH_PLACE, 1);
        WinningResult winningResult = new WinningResult(rankMap);

        //when
        double profitRateResult = winningResult.profitRate(1000);

        //then
        assertThat(profitRateResult).isEqualTo(5);
    }
```

leeyohan93 commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,70 @@
+        return new WinningResult(winningResult);
+    }
+
+    //TODO : 질문4
+    private void winningNumberSizeValid(List<Integer> winningNumbers) {
+        if (winningNumbers.size() != LottoProperty.LOTTO_COUNT) {
+            throw new IllegalArgumentException(ErrorMessage.LOTTO_WINNING_NUMBER_COUNT.getMessage());
+        }
+    }
```
>당첨번호의 개수를 검증하는 동일한 로직이 view와 domain에 들어가 있습니다.
view에서 검증을 하기 때문에 domain에서 빼는 것이 좋을까요? 아니면 중복이 되더라도 양쪽에서 진행을 하는 것이 좋을까요?

검증에 대해서는 View에서 검증을 해줄 수도 있습니다.
하지만 항상 저희는 잘못된 값이 넘어올 경우를 생각하고 검증해야합니다.
지금은 콘솔로 저희가 화면을 제어하지만 만약 로또 프로젝트가 웹으로 전환되었을 때를 생각해보시면 좋을 것 같습니다.

hongji3354 replied:
웹으로 전환한다 할 때 사용자로부터 넘겨받은 값을 Presentation layer에서 데이터 바인딩시 검증을 하고 문제가 없으면 Domain Layer에서는 이미 검증된 값이니 생각하고 사용하면 될 것이라 생각했습니다. 

그래도 Domain Layer에서도 한 번 더 검증을 해야하지 않을까 라고 생각하게 된 이유는 Domain Layer가 Application에서 가장 핵심이기 때문입니다.

근데 이렇게 된다면 로직이 중복이 될 것이고 만약 Lotto 번호의 개수가 변한다면 {resentation layer와 Domain Layer가 동시에 변경되어야 한다는 문제가 있기 때문에 고민을 하게 됐습니다.ㅎㅎ

leeyohan93 replied:
> Domain Layer가 Application에서 가장 핵심이기 때문입니다.

맞습니다. 도메인 레이어는 비즈니스의 핵심 로직이 구현되기 때문에 항상 신뢰성이 매우 중요합니다.
만약 resentation layer의 하나가 아닌 여러 개의 controller가 하나의 domain을 사용한다고 할 때, 어떤 controller는 검증을 놓칠 수도 있지 않을까요?  
개인적으로는 도메인에는 필수적으로 검증하는 것을 선호하는 것 같습니다 :)

leeyohan93 commented on README.md:
```diff
@@ -1,9 +1,34 @@
+
+## 3단계 - 로또(자동) 기능 요구사항 정리
+
+### Domain
```
기능 요구사항 잘 작성해주셨습니다 ! 👍

이에 더해 요구사항에 도메인 모델들을 도출하는 단계까지 하실 수 있을 것 같아요 :)

코드 구현 이전에 도메인 모델을 도출하고 구현해야할 기능들을 책임에 맞게 할당해보는 것도 또 다른 좋은 설계 방법인 것 같아요 !

(포비의 말씀처럼 도메인에 대한 이해가 없다면 일단 코드 구현! 그리고 코드 구현을 모두 날리고 다시 시작하는 방법도 좋습니다)

hongji3354 replied:
두 번정도 날리고 다시 구현을 했을 때 확실히 도메인을 도출하는게 훨씬 용이해 졌습니다!

leeyohan93 commented on src/main/java/lotto/view/LottoPurchaseView.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static int input() {
+        System.out.println("구입금액을 입력해주세요.");
+        Scanner scanner = new Scanner(System.in);
```
InputView가 로또 금액 뷰와 우승로또에 관련된 뷰로 두 가지로 나누어지면서 Scanner가 여러번 생성되고 있네요.
해당 부분을 성능을 최적화할 수 있는 방법은 없을까요 ? :)

leeyohan93 commented on src/main/java/lotto/view/LottoPurchaseView.java:
```diff
@@ -0,0 +1,38 @@
+    private static int numberValid(String purchaseAmount) {
+        try {
+            return Integer.parseInt(purchaseAmount);
+        } catch (NumberFormatException e) {
+            throw new IllegalArgumentException("구입금액은 숫자만 입력이 가능합니다.");
```
NumberFormatException은 IllegalArgumentException를 상속한 예외인데요!
에외의 추상화 레벨을 다시 올리신 이유가 있을까요 ?

hongji3354 replied:
catch 블럭이 NumberFormatException을 사용한 이유는 Integer.parseInt에 해당 예외를 던진다고 되어 있어서 사용하였으며 IllegalArgumentException으로 throw한 이유는 NumberFormatException은 숫자 형식과 관련된 예외로만 보이지만 IllegalArgumentException는 입력값 자체가 잘못되었구나 라고 생각하지 않을까 해서 입니다!

leeyohan93 replied:
NumberFormatException은 IllegalArgumentException를 상속하고 있기 때문에, 
`입력값 자체가 잘못되었구나` 라는 의미는 이미 내포하고 있다고 생각이 드는데요 :)

아래 링크에서 해당 아이템을 보시고 커스텀한 Exception을 만들어보는 것도 좋은 경험이 될 것 같아요 😄 

> 이펙티브 자바 아이템 73. 추상화 수준에 맞는 예외를 던지라
 https://madplay.github.io/post/effectivejava-chapter10-exceptions#%EC%95%84%EC%9D%B4%ED%85%9C-73-%EC%B6%94%EC%83%81%ED%99%94-%EC%88%98%EC%A4%80%EC%97%90-%EB%A7%9E%EB%8A%94-%EC%98%88%EC%99%B8%EB%A5%BC-%EB%8D%98%EC%A7%80%EB%9D%BC

leeyohan93 commented on src/main/java/lotto/domain/LottoStore.java:
```diff
@@ -0,0 +1,14 @@
+package lotto.domain;
+
+//구매 금액만큼 로또를 발급하는 역할을 가진 클래스
+public class LottoStore {
```
로또 발급에 대한 책임을 분리하셨네요 💯

leeyohan93 commented on src/main/java/lotto/domain/LottoIssuanceCount.java:
```diff
@@ -0,0 +1,51 @@
+        this.issuanceCount = issuanceCount;
+    }
+
+    public static LottoIssuanceCount issuanceNumberCalculation(int purchaseAmount) {
```
purchaseAmount 원시 값을 포장하는 Money와 같은 클래스를 도출해보는 것은 어떨까요 ? 
Money 클래스는 어떤 적합한 책임을 할당할 수 있을까요 ? :)

> 규칙 3: 모든 원시값과 문자열을 포장한다.

leeyohan93 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+// 로또 번호를 발급하는 역할을 가진 클래스
+public class LottoNumber {
```
로또 번호를 발급하는 책임이 너무 많은 클래스에 주어진 것이 아닐까요?
LottoStore 하나로 충분하다는 생각이 들기도 하는데 지훈님 의견은 어떠신가요?

클래스를 작게 분리하는 것은 정말 중요하지만, 적절한 책임에의해 만들어져야하며 항상 오버엔지니어링은 경계해야합니다. :)

hongji3354 replied:
엇 혹시 로또 번호를 발급하는 책임이 어느 클래스에도 분산되어 있는건가요?!

leeyohan93 replied:
피드백 드릴때 단어 선택이 잘못되었네요! ㅎㅎ 
로또 번호를 발급하는 것과 로또를 발급하는 책임을 나눈 것에 대한 이야기입니다 ! :)

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,79 @@
+// 로또
+public class Lotto {
+
+    private List<Integer> numbers;
```
로또 번호 숫자 하나를 의미하는 클래스를 도출해보는 것은 어떨까요 ?
이미 사용되고 있는 LottoNumber가 적절해 보이기도하는데.. 고민해볼 포인트인 것 같아요 🤔 

LottoNumber클래스를 사용하게되면 1~45 사이의 숫자라는 것을 믿고 신뢰하고 사용할 수 있습니다
아래 numberRangeValid() 검증 또한 Lotto에서 하지 않아도 되지요 :)

추가적으로 일급 컬렉션을 사용하실때는 final을 사용하면 좀 재할당을 막아 좀 더 불변에 가깝게 사용할 수 있습니다.

```suggestion
    private final List<Integer> numbers;
```

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,79 @@
+        numberCountValid(lottoNumbers);
+        numberDuplicateValid(lottoNumbers);
+        numberRangeValid(lottoNumbers);
+        this.numbers = lottoNumbers;
```
Collection에서 제공하는 unmodifiableList를 활용해보는 건 어떨까요?
리스트에 대한 추가, 삭제를 막아주는 기능으로 일급 컬렉션도 좀 더 불변에 가까워질 수 있습니다 :)
```suggestion
        this.numbers = Collections.unmodifiableList(lottoNumbers);
```

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,79 @@
+public class Lotto {
+
+    private List<Integer> numbers;
+    private int winningNumberMatchCount;
```
해당 필드는 꼭 상태로 가지고 있어야할까요?
꼭 필요한 필드만 캡슐화해주시는 것이 좋습니다 👍

hongji3354 replied:
테스트로 인해 사용한 필드인데, 다른 방법으로 테스트를 하여 제거 하였습니다!

leeyohan93 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,79 @@
+        this.winningNumberMatchCount = winningNumberMatchCount;
+    }
+
+    public int winningNumberMatchCount(List<Integer> winningNumbers) {
```
winningNumbers도 Lotto 클래스로 사용해볼 수 있지 않을까요 ? :)

hongji3354 replied:
앗 제가 이해를 못해서 그런데 더 상세히 설명해 주실수 있으실 까요?

leeyohan93 replied:
예시 코드로 보여드릴게요 :)

```suggestion
    public int winningNumberMatchCount(Lotto winningLotto) {
```

hongji3354 replied:
winningNumber를 Wrapping 한 클래스가 아닌 Lotto를 파라미터로 변경하게 되면 어떤 이점이 있는지 잘 모르겠습니다. 😢

leeyohan93 replied:
> 규칙 3: 모든 원시값과 문자열을 포장한다.

이 부분은 객체지향 생활 체조 원칙에 대한 부분이기도 합니다 :)
Lotto 클래스로 래핑하게 되면 Lotto 클래스에 걸어둔 아래 제약 조건 모두 사용 가능하게 되는데요 :)
지금의 List<Integer> 라는 파라미터 타입은 아래 두가지 사항에 대해 모두 신뢰할 수 없지 않을까요 ?

- 로또 번호는 1~45 사이의 중복되지 않은 숫자이다
- Lotto는 6자리 로또  번호이다

hongji3354 replied:
Lotto 클래스로 래핑시 Lotto 클래스에 걸어둔 제약 조건을 모두 사용하게 된다는 것은 엄청 큰 장점인 것 같습니다.
아니면 List<Integer> winningNumbers를 가지고 있는 일급 컬렉션을 따로 만드는 것도 괜찮을 것 같습니다.
다만, 제약 조건에 대한 중복 코드가 발생하기 때문에 트레이드 오프를 고려해 봐야 할 것 같습니다.

현재는 Lotto로 래핑하는 것이 훨씬 좋아보이네요!

leeyohan93 commented on src/test/java/lotto/domain/WinningTest.java:
```diff
@@ -0,0 +1,92 @@
+        );
+    }
+
+    @DisplayName("당첨 번호 일치 계산")
+    @ParameterizedTest
+    @MethodSource("matchCalculationParametersProvider")
+    void matchCalculation(List<Integer> lottoNumber, List<Integer> winningNumber) {
+        //given
+        Lotto lotto = new Lotto(lottoNumber);
+
+        //when
+        int winningNumberMatchResult = lotto.winningNumberMatchCount(winningNumber);
+
+        //then
+        assertThat(lotto).isEqualTo(new Lotto(lottoNumber, winningNumberMatchResult));
+    }
```
해당 테스트는 LottoTest에 좀 더 적합하지 않을까요? 🤔

leeyohan93 commented on src/main/java/lotto/domain/LottoProperty.java:
```diff
@@ -0,0 +1,14 @@
+package lotto.domain;
+
+public class LottoProperty {
```
저는 상수를 관리할 때, 따로 상수 전용 클래스를 선언하지 않고 상수가 사용되는 가장 가까운 곳에 배치하여 사용하고 있습니다.
보통 상수가 사용되는 가장 가까운 곳이라 하면 상수와 비즈니스적으로 밀접한 관련이 있는 클래스이며,
만약 public으로 여러 곳에서 참조해야할 경우는 가장 관련 있는 클래스 내부에 두고 관리하고 있습니다 :)

상수를 모아두는 클래스를 만드는 것 보다는 좀 더 관련있는 클래스 내부에 배치해주시는 것은 어떨까요?

hongji3354 replied:
상수가 사용되는 클래스에서 배치해서 사용을 했는데 public으로 변경하면서 상수 전용 클래스로 분리했습니다.

public 이더라도 해당 상수와 비즈니스적으로 밀접한 클래스에 위치하면 의미가 훨씬 명확해 질 것 같습니다!

leeyohan93 commented on src/main/java/lotto/view/LottoPurchaseListView.java:
```diff
@@ -0,0 +1,16 @@
+        System.out.print(lottos.purchaseNumber() + "개를 구매했습니다.");
+        System.out.println();
+
+        lottos.purchaseList().stream().forEach(System.out::println);
```
foreach 사용 시에는 stream을 사용하지 않으셔도 괜찮습니다 :)

```suggestion
        lottos.purchaseList().forEach(System.out::println);
```

leeyohan93 commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,70 @@
+        return this.lottos.size();
+    }
+
+    public List<Lotto> purchaseList() {
+        return this.lottos
+                .stream()
+                .collect(Collectors.toList());
+    }
```
방어적 복사를 사용하셨네요! 💯 
```suggestion
    public List<Lotto> purchaseList() {
        return new ArrayList<>(this.lottos);
    }
```

leeyohan93 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public enum Rank {
```
로또 당첨 결과를 의미하는 Rank 클래스 도출하셨네요 💯
Rank 클래스에 대한 테스트도 작성해주시면 좋을 것 같아요 :)

leeyohan93 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,48 @@
+        return rank;
+    }
+
+    public static boolean isPrize(int winningNumberMatchCount) {
+        return winningNumberMatchCount >= FOURTH_PLACE.getWinningNumberCount() && winningNumberMatchCount <= FIRST_PLACE.getWinningNumberCount();
+    }
```
요구사항이 복잡하게 변하여 이런 예외 케이스들이 점점 늘어나게 되면 Rank 클래스는 생각보다 더 복잡해질 가능성이 있습니다 :)
당첨되지 않았을 때의 Enum 타입을 추가해보는건 어떨까요?

leeyohan93 replied:
Rank.of 메소드를 재활용 하는 것으로는 상태를 확인하는 IsPrize라는 메서드를 제거할 수는 없을 것 같아요 🤔 
객체지향 프로그래밍의 원칙 중 디미터의 법칙이 있는데 이 링크가 도움이 되었으면 좋겠네요 :)

https://mangkyu.tistory.com/147

hongji3354 replied:
당첨되지 않았을 때 Enum 타입을 사용하는 것 보다 Rank.of 메소드를 재활용 하는 것은 어떨까요?

hongji3354 replied:
isPrize는 Rank에 있는 데이터를 외부에 노출시키지 않고 메시지를 주고 받은 것이라 생각해서 디미터의 법칙을 준수 했다고 생각하는데 어떤 부분 때문에 위반 인지 말씀해 주실 수 있으실까요?

hongji3354 replied:
> 요구사항이 복잡하게 변하여 이런 예외 케이스들이 점점 늘어나게 되면 Rank 클래스는 생각보다 더 복잡해질 가능성이 있습니다 :) 당첨되지 않았을 때의 Enum 타입을 추가해보는건 어떨까요?

Rank 클래스가 어떤 케이스로 인해 복잡해 질 수 있는지 잘 모르겠습니다ㅠㅠ
그래서 복잡해 졌을 때 isPrize가 왜 문제가 될 수 있는지를 잘 모르겠습니다. 🤔

leeyohan93 replied:
> isPrize는 Rank에 있는 데이터를 외부에 노출시키지 않고 메시지를 주고 받은 것이라 생각해서 디미터의 법칙을 준수 했다고 생각하는데 어떤 부분 때문에 위반 인지 말씀해 주실 수 있으실까요?

제가 설명이 많이 부족했네요. 
디미터의 법칙을 다른 말로 `묻지 말고 시켜라`원칙이라고 부르기도 하는데요.
데이터 자체를 외부에 노출 시키는 것이 아니라 메서드에 파라미터를 사용하도록 하는 것 또한 외부에 데이터를 노출 시키는 것이라고 볼 수 있습니다. 현재 상황에 빗대어 보면 로또의 match 수를 파라미터로 isPrize를 통해 Rank에게 당첨이 되었냐고 묻고있는 것이지요.저는 이 구조를 `Rank에게 당첨여부를 묻지 않아도 될 수 있는 구조로 개선`해보는 것은 어떨까요? 라는 코멘트를 드린 것이었습니다 😅 

사실 지금 요구사항에서 Rank가 더욱 복잡해지기는 힘들다고 생각합니다.
다만, 제가 복잡해진다고 말씀드렸던 부분은 객체에게 `묻지말고 시켜라`라는 원칙이 지켜지지 않는다면 다른 요구사항들이 나왔을 때 상태를 확인하는 메서드들이 언제든 추가될 수 있기 때문이라고 생각했기 때문입니다 :)

디미터 법칙은 캡슐화를 다른 관점에서 표현한 것이라고도 하는데요, 결과적으로는 디미터 법칙을 지키면서 캡슐화를 통해 좀 더 퍼블릭 인터페이스의 품질을 향상시킬 수 있다고 생각합니다 !

추가적으로 더 궁금하시면 언제든지 코멘트나 DM 남겨주세요. 제 의도가 잘 전달 되었나모르겠네요 😢

leeyohan93 replied:
> 조건문을 사용하는 것은 협력의 설계 측면에서 대부분의 경우 좋지 않은 선택이다. 항상 예외케이스를 최소화하고 일관성을 유지할 수 있는 방법을 선택하라.

설계를 위한 이런 가이드도 이해에 도움이 되었으면 합니다 👍

hongji3354 replied:
말해주신 부분 https://github.com/next-step/java-lotto-pro/pull/173/commits/36e2228104e0ece85604ce86d0bad7561da02d01 커밋에 반영했습니다.

말해주신 것과 일치하는 지 모르겠지만 isPrize 메서드의 파라미터가 없어지니 훨씬 좋다고 생각됩니다.

leeyohan93 reviewed (changes requested):
안녕하세요 지훈님 😄 

미션을 진행하면서 많이 고민한 흔적이 코드에서 보입니다 👍 👍 
몇가지 코멘트를 남겨드렸으니 확인부탁드려요 :)
제가 답변드린 사항에 대해 궁금하시면 언제든 코멘트, DM 남겨주세요 😄

hongji3354 committed:
refactor : 구매금액만큼 로또를 발급하는 기능 분리

Getter를 사용하지 않기 위해 로또발급개수 데이터를 가지고 있는 클래스에서 로또를 발급하는 기능을 제공하였으나 클래스가 가진 책임에 맞지도 않고 책임이 과도하다 생각하여 Getter를 제공하고 로또를 발급하는 책임을 가진 클래스에서 로또를 발급하는 기능을 제공하도록 변경

hongji3354 committed:
test : 구입 금액을 가지고 로또 발급 개수 확인하는 테스트

hongji3354 committed:
refactor : 한 곳에 모여있던 상수를 비즈니스적으로 밀접한 관련이 있는 클래스로 이동

hongji3354 committed:
refactor : foreach 사용 시 stream은 생략이 가능하므로 제거

hongji3354 committed:
refactor : 로또 구매 목록 제공하는 기능에서 생성자를 사용해서 return 하도록 변경

hongji3354 committed:
refactor : 당첨여부 확인하는 기능 내부 구현 수정

기존 메소드를 재사용하도록 변경

hongji3354 committed:
test : 로또 당첨 결과를 의미하는 Rank 테스트

hongji3354 committed:
test : 수익률 계산 테스트

hongji3354 committed:
refactor : 로또 번호를 가지고 있는 Collection 불변으로 변경

hongji3354 committed:
refactor : 로또 번호 숫자 하나를 클래스로 Wrapping

hongji3354 committed:
refactor : 로또 구매금액 클래스로 Wrapping

hongji3354 committed:
refactor : 당첨 번호와 일치 개수를 저장하는 변수 제거

hongji3354 committed:
refactor : 당첨 번호 일치 계산 테스트 클래스 이동

hongji3354 committed:
refactor : Scanner 캐싱해서 사용하도록 변경

hongji3354 committed:
fix : 당첨 번호 validation이 항상 false를 return 하도록 되어 있어서 수정

hongji3354 committed:
refactor : 사용자가 로또 구매 금액을 잘못 입력시 발생시키는 예외 커스텀 예외로 변경

hongji3354 committed:
refactor : 당첨 여부 기능을 제공하는 기능

hongji3354 committed:
refactor : 당첨 번호 일치 여부 확인하는 기능의 파라미터를 일급 컬렉션으로 변경

leeyohan93 commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,69 @@
+        this.lottos = lottos;
+    }
+
+    public WinningResult winningResult(List<Integer> winningNumbers) {
```
https://github.com/next-step/java-lotto-pro/pull/173#discussion_r744281714
winningNumbers를 Lotto 타입으로 만들어 보시는 걸 깜빡한 하신 것 같아요 :)
다음 단계에서 반영해주시면 좋을 것 같습니다 👍

leeyohan93 commented on src/main/java/lotto/domain/Lottos.java:
```diff
@@ -0,0 +1,69 @@
+        for (Lotto lotto : lottos) {
+            int winningNumberMatchCount = lotto.winningNumberMatchCount(winningNumber);
+            Rank rank = Rank.of(winningNumberMatchCount);
+            if (rank.isPrize()) {
+                winningResultAccumulate(winningResult, winningNumberMatchCount);
+            }
```
> 인덴트를 1로 유지한다.

위의 요구사항도 함께 지킬 수 있도록 개선해보는 건 어떨까요 ? :)

leeyohan93 reviewed (approved):
피드백 잘 반영해주셨습니다! :)
쉽지 않은 개선이었다고 생각하는데 잘 해주셨네요 👍 
남은 단계도 화이팅입니다 🔥
